### DISCUSSION TITLE
Question about autoplay

### DISCUSSION THREAD

**Discussion by IlyaTaidi** [matteobruni/tsparticles#1815] [Q&A] [answered]
Autoplay does not seem to work on refresh, but only after the screen is resized.

Safari/Angular

Any help would be appreciated

**Comment by matteobruni** [comment]
Can you give me more details about it? The config can be useful but some steps to reproduce the issue are welcome too

**Reply by IlyaTaidi** [reply]
Heres the config, I have enabled autoPlay, but for some reason only when I resize the browser do the actual particles come up from the top& Bottom of the screen

``` json
{
    "autoPlay": true,
    "background": {
      "color": {
        "value": "#15171e"
      },
      "image": "",
      "position": "50% 50%",
      "repeat": "no-repeat",
      "size": "cover",
      "opacity": 1
    },
    "backgroundMode": {
      "enable": true,
      "zIndex": 0
    },
    "detectRetina": true,
    "fpsLimit": 60,
    "interactivity": {
      "detectsOn": "canvas",
      "events": {
        "onClick": {
          "enable": true,
          "mode": "push"
        },
        "resize": true
      },
      "modes": {
        "push": {
          "quantity": 4
        },
      }
    },
    "motion": {
      "disable": false,
      "reduce": {
        "factor": 1,
        "value": false
      }
    },
    "particles": {
      "bounce": {
        "horizontal": {
          "random": {
            "enable": false,
            "minimumValue": 0.1
          },
          "value": 1
        },
        "vertical": {
          "random": {
            "enable": false,
            "minimumValue": 0.1
          },
          "value": 1
        }
      },
      "collisions": {
        "bounce": {
          "horizontal": {
            "random": {
              "enable": false,
              "minimumValue": 0.1
            },
            "value": 1
          },
          "vertical": {
            "random": {
              "enable": false,
              "minimumValue": 0.1
            },
            "value": 1
          }
        },
        "enable": true,
        "mode": "bounce"
      },
      "color": {
        "value": "#4b5680",
        "animation": {
          "enable": true,
          "speed": 1,
          "sync": true
        }
      },
      "life": {
        "count": 0,
        "delay": {
          "random": {
            "enable": false,
            "minimumValue": 0
          },
          "value": 0,
          "sync": false
        },
        "duration": {
          "random": {
            "enable": false,
            "minimumValue": 0.0001
          },
          "value": 0,
          "sync": false
        }
      },
      "links": {
        "blink": true,
        "color": {
          "value": "#39254c"
        },
        "consent": true,
        "distance": 100,
        "enable": true,
        "frequency": 1,
        "opacity": 0.61,
        "shadow": {
          "blur": 5,
          "color": {
            "value": "#00ff00"
          },
          "enable": false
        },
        "triangles": {
          "enable": true,
          "frequency": 0.62,
          "color": {
            "value": "#4a3a56"
          }
        },
        "width": 1,
        "warp": false
      },
      "move": {
        "angle": {
          "offset": 45,
          "value": 90
        },
        "attract": {
          "enable": false,
          "rotate": {
            "x": 600,
            "y": 1200
          }
        },
        "direction": "none",
        "distance": 0,
        "enable": true,
        "gravity": {
          "acceleration": 9.81,
          "enable": false,
          "maxSpeed": 50
        },
        "noise": {
          "delay": {
            "random": {
              "enable": false,
              "minimumValue": 0
            },
            "value": 0
          },
          "enable": false
        },
        "outModes": {
          "default": "out",
          "bottom": "out",
          "left": "out",
          "right": "out",
          "top": "out"
        },
        "random": false,
        "size": false,
        "speed": 0.6,
        "straight": false,
        "trail": {
          "enable": false,
          "length": 10,
          "fillColor": {
            "value": "#000000"
          }
        },
        "vibrate": false,
        "warp": false
      },
      "number": {
        "density": {
          "enable": true,
          "area": 800,
          "factor": 1000
        },
        "limit": 0,
        "value": 80
      },
      "opacity": {
        "random": {
          "enable": false,
          "minimumValue": 0.1
        },
        "value": 0.5,
        "animation": {
          "enable": false,
          "minimumValue": 0.1,
          "speed": 2.98,
          "sync": false
        }
      },
      "reduceDuplicates": false,
      "rotate": {
        "random": {
          "enable": false,
          "minimumValue": 0
        },
        "value": 0,
        "animation": {
          "enable": false,
          "speed": 0,
          "sync": false
        },
        "direction": "clockwise",
        "path": false
      },
      "shadow": {
        "blur": 0,
        "color": {
          "value": "#000000"
        },
        "enable": false,
        "offset": {
          "x": 0,
          "y": 0
        }
      },
      "shape": {
        "options": {
          "polygon": {
            "sides": 5
          },
          "image": {
            "src": "https://cdn.matteobruni.it/images/particles/github.svg",
            "width": 100,
            "height": 100
          }
        },
        "type": "circle"
      },
      "size": {
        "random": {
          "enable": true,
          "minimumValue": 1
        },
        "value": 1,
        "animation": {
          "destroy": "min",
          "enable": false,
          "minimumValue": 0.1,
          "speed": 1,
          "startValue": "max",
          "sync": false
        }
      },
      "stroke": {
        "width": 0,
        "color": {
          "value": "",
        }
      },
      "twinkle": {
        "lines": {
          "enable": true,
          "frequency": 0.05,
          "opacity": 1
        },
        "particles": {
          "enable": false,
          "frequency": 0.05,
          "opacity": 1
        }
      }
    },
  }
```

**Reply by matteobruni** [reply]
Autoplay is active by default, there's no need to set it to `true`. Can you share a video with the issue? I don't understand what happens when you resize the window.

**Reply by IlyaTaidi** [reply]
Heres a recording. It also shows the flickering issue from the other post.

https://user-images.githubusercontent.com/65737081/125706430-ec58278f-5cc5-4857-86f4-dbe21059dd13.mov

**Reply by IlyaTaidi** [reply]
I didn't see anything in the docs about it, but just to be sure. I don't need to initialize the particles with an OnInit function or anything right? Just in case, this is the ts file for the page linked above 

import { Component, OnInit } from '@angular/core';
import { OwlOptions } from 'ngx-owl-carousel-2';
import { Container, ISourceOptions, Main } from 'tsparticles';
export interface PhotosApi {
  albumId?: number;
  id?: number;
  title?: string;
  url?: string;
  thumbnailUrl?: string;
}

@Component({
  selector: 'app-home',
  templateUrl: 'home.page.html',
  styleUrls: ['home.page.scss'],
})
export class HomePage implements OnInit {
  title = 'angularowlslider';
  apiData: PhotosApi;
  limit: number = 10; // <==== Edit this number to limit API results
  customOptions: OwlOptions = {
    loop: true,
    autoplay: true,
    center: true,
    dots: false,
    autoHeight: false,
    autoWidth: true,
    responsive: {
      0: {
        items: 1,
      },
      600: {
        items: 1,
      },
      1000: {
        items: 1,
      }
    }
  };

  options: ISourceOptions = {
    "background": {
      "color": {
        "value": "#15171e"
      },
      "image": "",
      "position": "50% 50%",
      "repeat": "no-repeat",
      "size": "cover",
      "opacity": 1
    },
    "backgroundMode": {
      "enable": true,
      "zIndex": 0
    },
    "detectRetina": true,
    "fpsLimit": 60,
    "interactivity": {
      "detectsOn": "canvas",
      "events": {
        "onClick": {
          "enable": true,
          "mode": "push"
        },
        "resize": true
      },
      "modes": {
        "push": {
          "quantity": 4
        },
      }
    },
    "motion": {
      "disable": false,
      "reduce": {
        "factor": 1,
        "value": false
      }
    },
    "particles": {
      "bounce": {
        "horizontal": {
          "random": {
            "enable": false,
            "minimumValue": 0.1
          },
          "value": 1
        },
        "vertical": {
          "random": {
            "enable": false,
            "minimumValue": 0.1
          },
          "value": 1
        }
      },
      "collisions": {
        "bounce": {
          "horizontal": {
            "random": {
              "enable": false,
              "minimumValue": 0.1
            },
            "value": 1
          },
          "vertical": {
            "random": {
              "enable": false,
              "minimumValue": 0.1
            },
            "value": 1
          }
        },
        "enable": true,
        "mode": "bounce"
      },
      "color": {
        "value": "#4b5680",
        "animation": {
          "enable": true,
          "speed": 1,
          "sync": true
        }
      },
      "life": {
        "count": 0,
        "delay": {
          "random": {
            "enable": false,
            "minimumValue": 0
          },
          "value": 0,
          "sync": false
        },
        "duration": {
          "random": {
            "enable": false,
            "minimumValue": 0.0001
          },
          "value": 0,
          "sync": false
        }
      },
      "links": {
        "blink": true,
        "color": {
          "value": "#39254c"
        },
        "consent": true,
        "distance": 100,
        "enable": true,
        "frequency": 1,
        "opacity": 0.61,
        "shadow": {
          "blur": 5,
          "color": {
            "value": "#00ff00"
          },
          "enable": false
        },
        "triangles": {
          "enable": true,
          "frequency": 0.62,
          "color": {
            "value": "#4a3a56"
          }
        },
        "width": 1,
        "warp": false
      },
      "move": {
        "angle": {
          "offset": 45,
          "value": 90
        },
        "attract": {
          "enable": false,
          "rotate": {
            "x": 600,
            "y": 1200
          }
        },
        "direction": "none",
        "distance": 0,
        "enable": true,
        "gravity": {
          "acceleration": 9.81,
          "enable": false,
          "maxSpeed": 50
        },
        "noise": {
          "delay": {
            "random": {
              "enable": false,
              "minimumValue": 0
            },
            "value": 0
          },
          "enable": false
        },
        "outModes": {
          "default": "out",
          "bottom": "out",
          "left": "out",
          "right": "out",
          "top": "out"
        },
        "random": false,
        "size": false,
        "speed": 1.5,
        "straight": false,
        "trail": {
          "enable": false,
          "length": 10,
          "fillColor": {
            "value": "#000000"
          }
        },
        "vibrate": false,
        "warp": false
      },
      "number": {
        "density": {
          "enable": true,
          "area": 800,
          "factor": 1000
        },
        "limit": 0,
        "value": 80
      },
      "opacity": {
        "random": {
          "enable": false,
          "minimumValue": 0.1
        },
        "value": 0.5,
        "animation": {
          "enable": false,
          "minimumValue": 0.1,
          "speed": 2.98,
          "sync": false
        }
      },
      "reduceDuplicates": false,
      "rotate": {
        "random": {
          "enable": false,
          "minimumValue": 0
        },
        "value": 0,
        "animation": {
          "enable": false,
          "speed": 0,
          "sync": false
        },
        "direction": "clockwise",
        "path": false
      },
      "shadow": {
        "blur": 0,
        "color": {
          "value": "#000000"
        },
        "enable": false,
        "offset": {
          "x": 0,
          "y": 0
        }
      },
      "shape": {
        "options": {
          "polygon": {
            "sides": 5
          },
          "image": {
            "src": "https://cdn.matteobruni.it/images/particles/github.svg",
            "width": 100,
            "height": 100
          }
        },
        "type": "circle"
      },
      "size": {
        "random": {
          "enable": true,
          "minimumValue": 1
        },
        "value": 1,
        "animation": {
          "destroy": "min",
          "enable": false,
          "minimumValue": 0.1,
          "speed": 1,
          "startValue": "max",
          "sync": false
        }
      },
      "stroke": {
        "width": 0,
        "color": {
          "value": "",
        }
      },
      "twinkle": {
        "lines": {
          "enable": true,
          "frequency": 0.05,
          "opacity": 1
        },
        "particles": {
          "enable": false,
          "frequency": 0.05,
          "opacity": 1
        }
      }
    },
  }

  constructor() { }

  ngOnInit() {
  }

  particlesLoaded(container: Container): void {
    console.log(container);
  }

  particlesInit(main: Main): void {
    console.log(main);
  }

}

**Comment by IlyaTaidi** [comment, marked as answer]
Remember when I said this was fixed when loading the json is as a separate file? I was wrong.

Ive been trying to use ts particles as a background for a single webpage and I CAN NOT figure out why sometimes it auto starts & populates and other times it needs the window to be resized and populates from the bottom left. 

It is VERY frustrating, as it seems to do it whenever it wants to, sometimes with the same code. I have no idea if this is something that would only happen on localhost, or what. 

TLDR: Autoplay on Angualr/Ionic is still not working.

**Reply by matteobruni** [reply]
A similar issue with ionic was discusses some time ago with another user here: https://github.com/ramith123/Ionic-Angular-tsparticle-App/issues/1

Try this solution, maybe it will solve the starting issues, it's not related to `autoPlay` option since it's `true` by default. I'm creating a ionic demo in the project so I can see if there are issues that are under my control or a workaround is still needed.

I don't use Ionic so I'm not an expert about it.

**Reply by matteobruni** [reply]
I created a sample in the `dev` branch to replicate the issue and it's confirmed that the canvas created at the startup has width and height `0` so when you resize the window it fixes itself.

The library on resize tries to keep particles in the same place but they are all at 0,0 so they appear all together from a corner (out mode settings).

You can see the sample here: https://github.com/matteobruni/tsparticles/tree/dev/demo/ionic

In the `tab1` the particles are configured and have a refresh after 500ms and they work as expected.

**Reply by IlyaTaidi** [reply]
I downloaded and served up the project. Its serving and working, but I'm getting an error on the html file, 

Type '{ fullScreen: { enable: boolean; }; background: { color: { value: string; }; }; fpsLimit: number; interactivity: { detectsOn: string; events: { onClick: { enable: boolean; mode: string; }; onHover: { enable: boolean; mode: string; }; resize: boolean; }; modes: { ...; }; }; particles: { ...; }; detectRetina: boolean; }' is not assignable to type 'RecursivePartial<IOptions>'.
  The types of 'interactivity.detectsOn' are incompatible between these types.
    Type 'string' is not assignable to type '"canvas" | "window" | InteractivityDetect | "parent"'.
https://stackoverflow.com/questions/65803509/error-ts2322-in-angular-with-error-type-string-is-not-assignable-to-type-ca

Any idea on what this is?
I ran into this same issue on my own project, and its 'fixed' by changing the options to isSourceOptions, doing so removed the error but introduces that canvas problem. 

For now I'll be letting that error sit. Since it still works.

**Reply by matteobruni** [reply]
I don't know what is the issue, the cast is just something for TypeScript compiler, it shouldn't affect JavaScript runtime.
